File: section3/section 3.1/inflate.py
"""
ID: shikha11
LANG: PYTHON3
TASK: inflate
"""

fin = open('inflate.in', 'r')
fout = open('inflate.out', 'w')

M, N = map(int, fin.readline().strip().split())

points = [0] * N
minutes = [0] * N
max_points = [0] * (M + 1)

for p in range(N):
    points[p], minutes[p] = map(int, fin.readline().strip().split())
    t = minutes[p]
    for time in range(t, M + 1):
        max_points[time] = max(max_points[time], max_points[time - t] + points[p])


fout.write(str(max_points[M]) + '\n')
fout.close()

File: section3/section 3.1/inflate.java
/*
ID: shikha11
LANG: JAVA
TASK: inflate
 */

import java.io.*;
import java.util.StringTokenizer;

public class inflate {
  public static void main(String[] args) throws IOException {
    BufferedReader fin = new BufferedReader(new FileReader("inflate.in"));
    PrintWriter fout = new PrintWriter(new FileWriter("inflate.out"));

    StringTokenizer st = new StringTokenizer(fin.readLine());
    int M = Integer.parseInt(st.nextToken());
    int N = Integer.parseInt(st.nextToken());

    int[] points = new int[N];
    int[] minutes = new int[N];

    for (int i = 0; i < N; i++) {
      st = new StringTokenizer(fin.readLine());
      points[i] = Integer.parseInt(st.nextToken());
      minutes[i] = Integer.parseInt(st.nextToken());
    }

    int[] maxPoints = new int[M + 1];

    for (int p = 0; p < N; p++) {
      int t = minutes[p];
      for (int time = t; time <= M; time++) {
        maxPoints[time] = Math.max(maxPoints[time], maxPoints[time - t] + points[p]);
      }
    }

    fout.println(maxPoints[M]);
    fout.close();
  }
}
